x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/snapshotlifecycle/SnapshotLifecycleService.java

```diff
@@ -0,0 +1,124 @@
+    private final ClusterService clusterService;
+    private final SnapshotLifecycleTask snapshotTask;
+    private final Map<String, SchedulerEngine.Job> scheduledTasks = ConcurrentCollections.newConcurrentMap();
+    private volatile boolean isMaster = false;
+
+    public SnapshotLifecycleService(Settings settings, Client client, ClusterService clusterService,
+                                    Clock clock) {
+        this.scheduler = new SchedulerEngine(settings, clock);
+        this.clusterService = clusterService;
+        this.snapshotTask = new SnapshotLifecycleTask(client);
+        clusterService.addLocalNodeMasterListener(this); // TODO: change this not to use 'this'
+        clusterService.addListener(this);
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (this.isMaster) {
+            // TODO: handle modified policies (currently they are ignored)
+            // TODO: handle deleted policies
+            scheduleSnapshotJobs(event.state());
+        }
+    }
+
+    @Override
+    public void onMaster() {
+        this.isMaster = true;
+        scheduler.register(snapshotTask);
+        scheduleSnapshotJobs(clusterService.state());
+    }
+
+    @Override
+    public void offMaster() {
+        this.isMaster = false;
+        scheduler.unregister(snapshotTask);
+        cancelSnapshotJobs();
+    }
+
+    /**
+     * Schedule all non-scheduled snapshot jobs contained in the cluster state
+     */
+    public void scheduleSnapshotJobs(final ClusterState state) {
+        SnapshotLifecycleMetadata snapMeta = state.metaData().custom(SnapshotLifecycleMetadata.TYPE);
+        if (snapMeta != null) {
+            snapMeta.getSnapshotConfigurations().values().forEach(this::maybeScheduleSnapshot);
+        }
+    }
+
+    /**
+     * Schedule the {@link SnapshotLifecyclePolicy} job if it does not already exist. If the job already
+     * exists it is not interfered with.
+     */
+    public void maybeScheduleSnapshot(final SnapshotLifecyclePolicyMetadata snapshotLifecyclePolicy) {
+        final String jobId = snapshotLifecyclePolicy.getPolicy().getId();
+        scheduledTasks.computeIfAbsent(jobId, id -> {
+            final SchedulerEngine.Job job = new SchedulerEngine.Job(jobId,
+                new CronSchedule(snapshotLifecyclePolicy.getPolicy().getSchedule()));
+            logger.info("scheduling snapshot lifecycle job [{}]", jobId);
+            scheduler.add(job);
+            return job;
+        });
+    }
+
+    /**
+     * Cancel all scheduled snapshot jobs
+     */
+    public void cancelSnapshotJobs() {
+        scheduler.scheduledJobIds().forEach(scheduler::remove);
+        scheduledTasks.clear();
+    }
+
+    /**
+     * Cancel the given snapshot lifecycle id
+     */
+    public void cancelScheduledSnapshot(final String snapshotLifecycleId) {
+        scheduledTasks.remove(snapshotLifecycleId);
+        scheduler.remove(snapshotLifecycleId);
```

Comment:
If we add the job using `computeIfAbsent` on a `ConcurrentHashMap`, we should probably remove it using `computeIfPresent`, e.g.:
```
scheduledTasks.computeIfPresent(snapshotLifecycleId, (id, job) -> {
  scheduler.remove(snapshotLifecycleId);
  return null;
});
```
That will make sure we have the same atomicity guarantees coming and going.

Reply:
The `remove` for the scheduler is idempotent so I erred on the side of making sure to absolute remove the job even if the task map was out of sync.

I fully expect the implementation for this in particular (the map of tasks) to change, since right now it doesn't handle updates :)